---
title: "Single Cell RNA-Seq"
author: "Bengy Gyimah-Asamoah"
date: "2023-12-13"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::knit_hooks$set(purl = knitr::hook_purl)
knitr::opts_chunk$set(echo = TRUE)

```

Workflow conducting a single-cell RNA-Seq analysis in R using Seurat on a 10x Genomics dataset containing sorted cells from human Glioblastoma Multiforme, a fast-growing brain tumor.

```{r Load libraries}
library(rmarkdown)
library(Seurat)
library(tidyverse)
library(dplyr)
library(hdf5r)
```

Single-cell data was obtained from the 10x Genomics database <https://www.10xgenomics.com/resources/datasets/2-k-sorted-cells-from-human-glioblastoma-multiforme-3-v-3-1-3-1-standard-6-0-0>

```{r Loading data}
# Loading Hodgkin's Lymphoma tumor dataset
braintumor.data<- Read10X_h5(filename = 'D:/My Drive/Projects/R/Single-Cell RNA-Seq/data/Brain_Tumor_3p_raw_feature_bc_matrix.h5')

# Initialize the Seurat object with the raw (non-normalized data)
braintumor <- CreateSeuratObject(counts = braintumor.data, project = "Hodgkin's Lymphoma", min.cells = 3, min.features = 200)
braintumor

```

Typical QC controls for low-quality cells or empty droplets that often have very few genes, cell doublets or multiplets that may exhibit an aberrantly high gene count, and low-quality / dying cells often exhibit extensive mitochondrial contamination

Mitochondrial QC metrics will be calculated with the PercentageFeatureSet() function, which calculates the percentage of counts originating from a set of features

```{r QC}
#Mitochondrial QC metrics
braintumor[["percent.mt"]] <- PercentageFeatureSet(braintumor, pattern = "^MT-")

```

Creating visualizations of QC metrics to determine filter parameters

```{r Plots}
# Violin plot
VlnPlot(braintumor, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
view(braintumor@meta.data)

# Scatter plot to visualize feature-feature relationships
plot1 <- FeatureScatter(braintumor, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(braintumor, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1
plot2

```

Based on the data from the plots above we can filter for unique feature counts over 7500, unique feature counts less than 200, and \>20% mitochondrial counts

```{r Filtering}
braintumor <- subset(braintumor, subset = nFeature_RNA > 200 & nFeature_RNA < 7500 & percent.mt < 20)

```

```{r Normalization}
# Normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.
braintumor <- NormalizeData(braintumor)

```

Feature selection involves the calculation of features that exhibit high cell-to-cell variation in the dataset, those that are highly expressed in some cells, and lowly expressed in others by directly modeling the mean-variance relationship inherent in single-cell data

```{r Feature selection}
# Identifying variable features
braintumor <- FindVariableFeatures(braintumor, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(braintumor), 10)

# plot variable features with and without labels
plot3 <- VariableFeaturePlot(braintumor)
plot4 <- LabelPoints(plot = plot3, points = top10, repel = T)
plot3
plot4
```

Scaling the data involves linear transformation which is a standard pre-processing step prior to dimensional reduction techniques like PCA. This shifts the expression of each gene so that the mean expression across cells is 0 then scales the expression of each gene, so that the variance across cells is 1. This step gives equal weight in downstream analyses so that highly-expressed genes do not dominate

```{r Scaling}
all.genes <- rownames(braintumor)
braintumor <- ScaleData(braintumor, features = all.genes)
```

Principal component analysis (PCA) involves dimensionality reduction to identify sources of heterogeneity in the dataset

```{r PCA}
braintumor <- RunPCA(braintumor, features = VariableFeatures(object = braintumor))

# Checking if cells cluster in PCA
DimPlot(braintumor, reduction = "pca") + NoLegend()
```

DimHeatmap allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses

```{r Heatmap}
# Both cells and features are ordered according to their PCA scores
DimHeatmap(braintumor, dims = 1:3, cells = 500, balanced = T)
```

To properly determine the dimensionality of the dataset an elbow plot will be used to visualize the rankings of the principle components

```{r Elbow plot}
# Elbow plot: a ranking of principle components based on the percentage of variance explained by each one 
ElbowPlot(braintumor)
```

Clustering involves creating clusters of similar cells which have similar feature expression patterns

```{r Clustering}
braintumor <- FindNeighbors(braintumor, dims = 1:15)
braintumor <- FindClusters(braintumor, resolution = c(0.1,0.3, 0.5, 0.7, 1))

# Visualization
DimPlot(braintumor, group.by = "RNA_snn_res.0.1", label = T)

# Setting identity of clusters
Idents(braintumor) <- "RNA_snn_res.0.1"

```

```{r Non-linear dimensionality reduciton}
braintumor <- RunUMAP(braintumor, dims = 1:15)
DimPlot(braintumor, reduction = "umap")
```

Finding markers specific to each cluster

```{r Specific clusters}
# Finding all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(braintumor, ident.1 = 5, ident.2 = c(0, 3))
head(cluster5.markers, n = 5)

# Finding markers for every cluster compared to all remaining cells, reporting only the significant ones
braintumor.markers <- FindAllMarkers(braintumor, only.pos = T)
braintumor.markers %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1)

# Plotting raw counts
VlnPlot(braintumor, features =c("NOVA1", "CD14", "LYZ", "NKG7") , slot = "counts", log = T)

FeaturePlot(braintumor, features = c("CD14", "NOVA1", "FCGR3A", "LYZ", "CD8A", "IL7R"))
```

Assigning cell type identity to clusters using canonical markers to match the unbiased clustering to known cell types

```{r Cell typing}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono")
names(new.cluster.ids) <- levels(braintumor)
braintumor <- RenameIdents(braintumor, new.cluster.ids)
DimPlot(braintumor, reduction = "umap", label = T, pt.size = 0.5) + NoLegend()
```

